contracts/TapETH.sol

```diff
@@ -0,0 +1,545 @@
+     */
+    function sharesOf(address _account) external view returns (uint256) {
+        return _sharesOf(_account);
+    }
+
+    /**
+     * @return the amount of shares that corresponds to `_tapETHAmount` protocol-controlled tapETH.
+     */
+    function getSharesByPooledEth(
+        uint256 _tapETHAmount
+    ) public view returns (uint256) {
+        uint256 _totalPooledEther = _getTotalPooledEther();
+        if (_totalPooledEther == 0) {
+            return 0;
+        } else {
+            return (_tapETHAmount * _getTotalShares()) / _totalPooledEther;
+        }
+    }
+
+    /**
+     * @return the amount of tapETH that corresponds to `_sharesAmount` token shares.
+     */
+    function getPooledEthByShares(
+        uint256 _sharesAmount
+    ) public view returns (uint256) {
+        if (totalShares == 0) {
+            return 0;
+        } else {
+            return (_sharesAmount * _totalSupply) / (totalShares);
+        }
+    }
+
+    /**
+     * @notice Moves `_sharesAmount` token shares from the caller's account to the `_recipient` account.
+     * @return amount of transferred tokens.
+     * Emits a `TransferShares` event.
+     * Emits a `Transfer` event.
+     * @dev The `_sharesAmount` argument is the amount of shares, not tokens.
+     */
+    function transferShares(
+        address _recipient,
+        uint256 _sharesAmount
+    ) external returns (uint256) {
+        _transferShares(msg.sender, _recipient, _sharesAmount);
+        uint256 tokensAmount = getPooledEthByShares(_sharesAmount);
+        _emitTransferEvents(
+            msg.sender,
+            _recipient,
+            tokensAmount,
+            _sharesAmount
+        );
+        return tokensAmount;
+    }
+
+    /**
+     * @notice Moves `_sharesAmount` token shares from the `_sender` account to the `_recipient` account.
+     *
+     * @return amount of transferred tokens.
+     * Emits a `TransferShares` event.
+     * Emits a `Transfer` event.
+     *
+     * Requirements:
+     * - the caller must have allowance for `_sender`'s tokens of at least `getPooledEthByShares(_sharesAmount)`.
+     * @dev The `_sharesAmount` argument is the amount of shares, not tokens.
+     */
+    function transferSharesFrom(
+        address _sender,
+        address _recipient,
+        uint256 _sharesAmount
+    ) external returns (uint256) {
+        uint256 tokensAmount = getPooledEthByShares(_sharesAmount);
+        _spendAllowance(_sender, msg.sender, tokensAmount);
+        _transferShares(_sender, _recipient, _sharesAmount);
+        _emitTransferEvents(_sender, _recipient, tokensAmount, _sharesAmount);
+        return tokensAmount;
+    }
+
+    function mintShares(address _account, uint256 _tokenAmount) external {
+        require(pools[msg.sender], "TapETH: no pool");
+        _mintShares(_account, _tokenAmount);
+    }
+
+    function burnShares(uint256 _tokenAmount) external {
+        _burnShares(msg.sender, _tokenAmount);
```

Comment:
not sure if we want to use the shares or token amount as the input argument here.